talk/morelanguage/raii.tex

```diff
@@ -256,6 +288,125 @@
   \end{block}
 \end{frame}
 
+\begin{frame}[fragile]
+  \frametitlecpp[11]{weak\_ptr}
+  \begin{block}{weak\_ptr: a non-owning observer}
+    \small
+    \begin{itemize}
+    \item Sometimes want to observe resources without keeping them alive
+    \begin{itemize}
+      \item \mintinline{cpp}{shared_ptr}? Resource stays alive
+      \item Raw pointer? Risk of dangling pointer
+    \end{itemize}
+    \item The solution is to construct a \mintinline{cpp}{weak_ptr} from a shared pointer
+    \item To access the resource, convert the weak into a \mintinline{cpp}{shared_ptr}
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{}
+    \small
+    \begin{cppcode*}{}
+      std::shared_ptr<Cache> getSharedCache();
+      std::weak_ptr<Cache> weakPtr{ getSharedCache() };
+      // ... shared cache may be invalidated here
+      if (std::shared_ptr<Cache> cache = weakPtr.lock()) {
+        // Cache is alive, we actively extend its lifetime
+        return cache->findItem(...);
+      } else {
+        // Cache is nullptr, we need to do something
+        weakPtr = recomputeCache(...);
```

Comment:
This confuses me. What does `recomputeCache(...)` return? If it's a shared ptr, it would be immediately destroyed. And returning a weak ptr seems odd to me.